File: 2025-08-August-GFG-POTD/August 13 - Tywin's War Strategy/Solution.py
class Solution:
    def minSoldiers(self, arr, k):
        n = len(arr)
        required = (n + 1) // 2  # ceil(n / 2)
        lucky_count = 0
        add_list = []

        for num in arr:
            if num % k == 0:
                lucky_count += 1
            else:
                add_list.append(k - (num % k))

        if lucky_count >= required:
            return 0

        add_list.sort()
        to_make_lucky = required - lucky_count
        soldiers_needed = sum(add_list[:to_make_lucky])

        return soldiers_needed


# Example usage
sol = Solution()
print(sol.minSoldiers([3, 5, 6, 7, 9, 10], 4))  # Output: 4

File: 2025-08-August-GFG-POTD/August 13 - Tywin's War Strategy/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minSoldiers(vector<int>& arr, int k) {
        int n = arr.size();
        int required = (n + 1) / 2; // ceil(n / 2)
        int luckyCount = 0;
        vector<int> addList;

        for (int num : arr) {
            if (num % k == 0) {
                luckyCount++;
            } else {
                addList.push_back(k - (num % k));
            }
        }

        if (luckyCount >= required) {
            return 0;
        }

        sort(addList.begin(), addList.end());

        int soldiersNeeded = 0;
        int toMakeLucky = required - luckyCount;

        for (int i = 0; i < toMakeLucky; i++) {
            soldiersNeeded += addList[i];
        }

        return soldiersNeeded;
    }
};

int main() {
    Solution sol;
    vector<int> arr = {3, 5, 6, 7, 9, 10};
    int k = 4;
    cout << sol.minSoldiers(arr, k) << endl; // Output: 4
    return 0;
}
